### DISCUSSION TITLE
tutorials/networking/high_level_multiplayer

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#57] [User-contributed notes]
# tutorials/networking/high_level_multiplayer

High-level vs low-level API: The following explains the differences of high- and low-level networking in Godot as well as some fundamentals. If you want to jump in head-first and add networking to ...

https://docs.godotengine.org/en/stable/tutorials/networking/high_level_multiplayer.html

**Comment by swAAn01** [comment]
I'm quite new to networking and multiplayer, so sorry if this question is common. If I'm making a multiplayer game, is my selection of the port to run it on completely arbitrary? I see that the port needs to be specified in many places, so can this just be any number?

**Reply by hawkgamingdev** [reply]
Exactly you choose what you want, you can also google the most recommended ports to use in your games.

**Reply by Calinou** [reply]
You can use any port between 1025 and 49150 that is not a "registered port" listed on [Wikipedia](https://en.wikipedia.org/wiki/List_of_TCP_and_UDP_port_numbers).

Ports 1-1024 should be avoided, as only the root user can listen on them on platforms other than Windows. This can complicate the server setup process, as the system administrator will need to explicitly grant capabilities for a limited to user to listen on a port below 1024.

Ports 49151-65535 should also be avoided as they are considered "ephemeral" ports (and are used by clients to connect to the server).

**Reply by 3Rton** [reply]
For whatever it's worth, recently I've been maintaining a V Rising server for friends and for example there you use a config file for the server to define which ports are used (the server maintainer) and on the connection side the clients must input which port they are trying to connect through.
I think that's pretty good approach considering you can't know the situation someone who wants to host a server might be in (though if you only ever plan to host a server yourself I guess you can hard code it)

**Comment by jjolmo** [comment]
I was looking for the parameters that an rpc accepts, then I see this:

```
RPCs will not serialize objects or callables.

```
The naming is a bit confusing, maybe it's just me but I don't even know what is serializing or if this serializing has  something to do to the parameters that a RPC accept as payload. Wouldn't make more sense and help users to say something like
RPCs will only accept simple types (int, float, string) as parameters?

**Reply by KyleLevi** [reply]
Serializing is a common low-level networking term. Usually, if you're sending data over a network, you need to turn it into something like bytes first (that is the process of serialization). 

The trouble is that the other computer needs to de-serialize the data back into something meaningful. In theory you could serialize almost anything, but in practice you usually just want to pass a reference to a synchronized object (so instead of sending an entire object over the network, you can pass a tiny reference to an object).  This is basically what the High-level API is doing for us behind the scenes.

In the future, there will probably be docs on how to write and implement your own serialization methods for your custom classes/data. But that's quite advanced... [as an example, I'm going to link to Unity's version of it - please don't hurt me! https://docs-multiplayer.unity3d.com/netcode/current/advanced-topics/custom-serialization/]

**Reply by salmonmoose** [reply]
The range of things that can be sent are broad enough to carry dictionaries, and packed byte arrays.

This makes serialization fairly straight forward if you don't go with a generic solution, and just use something bespoke.

Make a resource type and give it methods that generate a dictionary, and allow it to be built from a dictionary.

I'm using this to push around player data for example - also shows an example of pushing an image in the request.

```gdscript
extends Resource
class_name PlayerData

var name : String
var avatar : Image
var completed_levels : Dictionary = {}
var id : int
var online : bool

func serialize() -> Dictionary:
	return {
		'name': name,
		'avatar': avatar.data
	}

static func deserialize(data: Dictionary) -> PlayerData:
	var playerData : PlayerData = PlayerData.new()
	playerData.avatar = Image.create_from_data(
		int(data.avatar.width),
		int(data.avatar.height),
		false,
		Image.FORMAT_RGBA8,
		data.avatar.data as PackedByteArray
		)

	playerData.name = data.name
	
	return playerData
```

**Comment by RoboClonk** [comment]
| "authority": Only the multiplayer authority ~(the server)~ can call remotely.

This line is misleading. Only the authority may call remotely, which *can* be the server, but not necessarily.

**Reply by Calinou** [reply]
Pull request opened:

- https://github.com/godotengine/godot-docs/pull/10001

**Comment by BaconEggsRL** [comment]
Quick tip for anyone looking to make a browser-based game: You won't be able to use the same ENetMultiplayerPeer like what is in this tutorial. This is because ENet is just UDP behind the scenes, and browsers don't support UDP out of the box. For high-level multiplayer you'll either need to use WebSocket or WebRTC.

Some links for extra reading below:
https://docs.godotengine.org/en/stable/tutorials/export/exporting_for_web.html#networking
https://gafferongames.com/post/why_cant_i_send_udp_packets_from_a_browser/

**Reply by Neo-Ciber94** [reply]
Damn, I should have checked this thread before starting lol

I was confused because the game was not firing requests, time to refactor.

**Comment by oof-fr** [comment]
sorry if i seem dumb, but what node do i attach the lobby script to?

**Reply by Calinou** [reply]
In the [Multiplayer Bomber](https://github.com/godotengine/godot-demo-projects/tree/master/networking/multiplayer_bomber) demo, `lobby.gd` is attached to the root node of `lobby.tscn` (which is the main scene in that project).

**Comment by MichaelGumm** [comment]
Am I missing something or does the example code for the Lobby scene not work? I'm trying to follow along and am running into errors with "_register_player.rpc_d(id, player_info)" line not being able to be called on oneself in the "_on_player_connected" method. Not to mention this "_on_player_connected" method being declared with one variable but the signal calls to it all use two variables...

**Reply by midiphony** [reply]
It is a [known issue](https://github.com/godotengine/godot/issues/68407#issuecomment-2331790407) and the documentation should reflect it. We should change the example lobby implementation with one working in the current state of the engine, and add a disclaimer about the problematic Godot API in multiplayer.

There is a [related issue](https://github.com/godotengine/godot-docs/issues/10723) listed in the godot-docs repository. If I get the time, I might propose a working implementation based on the [workaround](https://godotengine.org/article/multiplayer-in-godot-4-0-scene-replication/#selecting-levels) shared by @Faless

**Comment by KainSlow** [comment]
For anyone having problems freeing the port that was used to create the server, you need to destroy the underlying [ENetConnection](https://docs.godotengine.org/en/stable/classes/class_enetconnection.html) after calling `remove_multiplayer_peer()` on the server, one way to do this is to hold a class reference of your `ENetMultiplayerPeer peer`, then after you disconnect the server, you simply call, `peer.host.destroy()`, otherwise, if you try to use the same port again, you will get the **cant_create** error.

I find it kinda weird that this isn't mentioned here or in any tutorial i have seen, not even on the [multiplayer bomber demo](https://godotengine.org/asset-library/asset/2797) found in the asset lib.

Note that you only need to do this server side.

**Reply by Calinou** [reply]
This seems to be a bug. Can you open an issue on https://github.com/godotengine/godot/issues with a minimal reproduction project attached?

**Reply by KainSlow** [reply]
Done: [Here](https://github.com/godotengine/godot/issues/102822) you go

**Reply by KainSlow** [reply]
Just so everyone knows, this issue got fixed as of Godot stable mono 4.4.1, but due to C# Garbage Collector limitations, we need to explicitly call `peer.Close()` or `peer.Dispose()` when closing the server if we want it to free the port immediately.
Note that this does not happen in GDScript, only in C#.
I think this should be mentioned in the docs as a warning, since this is more of a C# limitation that users should be aware of.

**Comment by crizmo** [comment]
How should i go about implementing a room based system where the clients can either join room 1 / room 2 and can only see , interact , chat with each other if they are in the same room with this ?
Thanks !

**Comment by noidexe** [comment]
WARNING: Do not use change_scene_to methods in multiplayer as shown in "Example lobby implementation". It can cause synchronization issues. 

See related bug report with a comment from Faless acknowledging the issue and linking to a workaround: https://github.com/godotengine/godot/issues/68407#issuecomment-2331790407

This is what's recommended instead: https://godotengine.org/article/multiplayer-in-godot-4-0-scene-replication/#selecting-levels

I'm opening an issue in godot-docs to fix this

**Comment by jak6jak** [comment]
I think the access modifiers may be incorrect for the C# implementation of the lobby. I could be wrong though.

**Comment by gtament** [comment]
Noticed a discrepancy in the documentation: the `multiplayer.multiplayer_peer` property's default value isn't mentioned, and it differs from what one might infer from the *"To terminate networking"* section. It would be helpful to explicitly state that the default is `OfflineMultiplayerPeer`, as this can lead to misunderstandings — for instance, all RPC calls will stop working in single-player mode after terminating a multiplayer session the suggested way. Clarifying this behaviour in the documentation would greatly improve developer experience and prevent confusion during implementation. Thanks!

**Comment by MichaelrMentele** [comment]
It would be great to describe a canonical way to get the server IP address for client connection for LAN.

I'm doing this atm, but seems like it's a very common issue to know what IP address you are actually running on so you can broadcast it to potential clients.

```gdscript
func get_lan_ip() -> String:
	for ip in IP.get_local_addresses():
		if ip.is_valid_ip_address() \
		and ip.find('.') != -1 \
		and not ip.begins_with("127.") \
		and not ip.begins_with("169.254."):
			return ip
	push_error("No usable LAN IP found.")
	return ""
```

**Comment by wolfsmith520-github** [comment]
I am currently working on a p2p multiplayer game, and I am trying to get everyone’s ping on screen, I use `get_statistic(ENetPacketPeer.PEER_LAST_ROUND_TRIP_TIME)` from `ENetPacketPeer`
But there’s one problem, currently I store player’s ENetPacketPeer to an autoload script when he joined the game, but these is no way a rpc function can get that. I am wondering if I can get someone’s ENetPacketPeer with the peer id. Or Is there a better solution to display everyone’s ping on screen?

**Reply by noidexe** [reply]
Yes. Make something like `@rpc("any_peer", "call_local", "unreliable") update_ping_for_peer( peer_id, ping )`. Then you can call that rpc from every player as often as you like (maybe once per second). Each player should be able to obtain it's own peer_id and statistics so you just share that over the network. Then inside the function with that info you have all you need to update the UI.

**Comment by jmhnilbog** [comment]
It looks to me like the "Example lobby implementation" example doesn't actually work.  Shouldn't something call Lobby::create_game()? Right now, a player gets loaded but nothing ever adds player info to the players object, and players.size() remains 0.

**Comment by mavilla-dev** [comment]
> It is possible to create a new MultiplayerAPI object and assign it to a NodePath in the scene tree, which will override multiplayer for the node at that path and all of its descendants. This allows sibling nodes to be configured with different peers, which makes it possible to run a server and a client simultaneously in one instance of Godot.

Docs mention running server and client within the same host but I couldn't easily find an example that did this so I wrote one up incase others had the same question. https://github.com/mavilla-dev/godot-mp-chat

**Comment by l-e-webb** [comment]
I found this manual incomplete without also reading the documentation for `MultiplayerSynchronizer` and `MultiplayerSpawner`:
https://docs.godotengine.org/en/stable/classes/class_multiplayersynchronizer.html
https://docs.godotengine.org/en/stable/classes/class_multiplayerspawner.html

As is, this documentation stops short of actually instruction on how to synchronize game state. The synchronizer and spawner nodes are complicated enough that I, for one, would find a high level overview of how they work and how to use them extremely helpful.
